pkg/spec/deployment_transform.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2017 The Kedge Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spec
+
+import "fmt"
+
+import (
+	log "github.com/Sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api_v1 "k8s.io/client-go/pkg/api/v1"
+	ext_v1beta1 "k8s.io/client-go/pkg/apis/extensions/v1beta1"
+
+	"reflect"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/pkg/api"
+)
+
+// Transform function if given DeploymentSpecMod data creates the versioned
+// kubernetes objects and returns them in list of runtime.Object
+// And if the field in DeploymentSpecMod called 'extraResources' is used
+// then it returns the filenames mentioned there as list of string
+func (deployment *DeploymentSpecMod) Transform() ([]runtime.Object, []string, error) {
+
+	runtimeObjects, extraResources, err := deployment.CreateK8sObjects()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to create Kubernetes objects")
+	}
+
+	deploy, err := deployment.CreateK8sController()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to create Kubernetes Deployment controller")
+	}
+
+	// adding controller objects
+	// deployment will be nil if no deployment is generated and no error occurs,
+	// so we only need to append this when a legit deployment resource is returned
+	if deploy != nil {
+		runtimeObjects = append(runtimeObjects, deploy)
+		log.Debugf("deployment: %s, deployment: %s\n", deploy.Name, spew.Sprint(deployment))
+
+	}
+
+	if len(runtimeObjects) == 0 {
+		return nil, nil, errors.New("No runtime objects created, possibly because not enough input data was passed")
+	}
+
+	for _, runtimeObject := range runtimeObjects {
+
+		gvk, isUnversioned, err := api.Scheme.ObjectKind(runtimeObject)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "ConvertToVersion failed")
+		}
+		if isUnversioned {
+			return nil, nil, fmt.Errorf("ConvertToVersion failed: can't output unversioned type: %T", runtimeObject)
+		}
+
+		runtimeObject.GetObjectKind().SetGroupVersionKind(gvk)
+	}
+
+	return runtimeObjects, extraResources, nil
+}
+
+// Creates a Deployment Kubernetes resource. The returned Deployment resource
+// will be nil if it could not be generated due to insufficient input data.
+func (deployment *DeploymentSpecMod) CreateK8sController() (*ext_v1beta1.Deployment, error) {
```

Comment:
Why is this called generically as `CreateK8sController` if its always returning Deployment?

Reply:
So, when we will have Jobs, we can have a separate method -
`func (job *JobSpecMod) CreateK8sController() (*batch.Job, error)`

Reply:
Do they have to have the same name?

Why it can't be `CreateDeployment` and `CreateJob`?

Reply:
Yeah, but since both of them are controllers, I thought having the same method on both would make sense. Would bring a bit of consistency in code.
What do we lost by this?